### PR TITLE
Handle platform specific files in plugin

### PR DESCRIPTION
**Pull request by Fatme**
Fixes https://github.com/NativeScript/nativescript-cli/issues/578

### CODE REVIEW

File: lib/services/project-files-manager.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+///<reference path="../.d.ts"/>
+"use strict";
+import path = require("path");
+import util = require("util");
+
+export class ProjectFilesManager implements IProjectFilesManager {
+	constructor(private $fs: IFileSystem,
+		private $platformsData: IPlatformsData) { }
+	
+	public processPlatformSpecificFiles(directoryPath: string, platform: string, excludedDirs?: string[]) {
```
Comment (teobugslayer) [inline lib/services/project-files-manager.ts]: please declare function return type

File: lib/services/project-files-manager.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		return (() => {
+			var contents = this.$fs.readDirectory(directoryPath).wait();
+			var files: string[] = [];
+			
+			_.each(contents, fileName => {
+				let filePath = path.join(directoryPath, fileName);
+				let fsStat = this.$fs.getFsStats(filePath).wait();
+				if(fsStat.isDirectory() && (!excludedDirs || (excludedDirs && !_.contains(excludedDirs, fileName)))) {
```
Comment (rosen-vladimirov) [inline lib/services/project-files-manager.ts]: you do not need the !excludeDirs check - contains returns false if the first parameter is null/undefined:
```
> l.contains(undefined, "test")
false
> l.contains(null, "test")
false
```

File: lib/services/project-files-manager.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+					this.processPlatformSpecificFilesCore(platform, this.$fs.enumerateFilesInDirectorySync(filePath)).wait();
+				} else if(fsStat.isFile()) {
+					files.push(filePath);
+				}
+			});
+			this.processPlatformSpecificFilesCore(platform, files).wait();
```
Comment (rosen-vladimirov) [inline lib/services/project-files-manager.ts]: Maybe we can improve the performance here:
```TypeScript
let futures: IFuture[];
_.each(contents, fileName => {
	let filePath = path.join(directoryPath, fileName);
	let fsStat = this.$fs.getFsStats(filePath).wait();
	if(fsStat.isDirectory() && !_.contains(excludedDirs, fileName)) {
		future.push(this.processPlatformSpecificFilesCore(platform, this.$fs.enumerateFilesInDirectorySync(filePath)));
	} else if (fsStat.isFile()) {
		files.push(filePath);
});
futures.push(this.processPlatformSpecificFilesCore(platform, files));
Future.wait(futures);
```
I've not tested it and this is **not** a merge stopper.

File: lib/services/project-files-manager.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+			
+		}).future<void>()();
+	}
+	
+	private processPlatformSpecificFilesCore(platform: string, files: string[]): IFuture<void> {
+		// Renames the files that have `platform` as substring and removes the files from other platform
+		return (() => {
+			_.each(files, fileName => {
```
Comment (rosen-vladimirov) [inline lib/services/project-files-manager.ts]: Also not a merge stopper, but consider:
```TypeScript
let futures = files.map(fileName => {
	let platformInfo = ProjectFilesManager.parsePlatformSpecificFileName(path.basename(fileName), this.$platformsData.platformsNames);
	let shouldExcludeFile = platformInfo && platformInfo.platform !== platform;
	if (shouldExcludeFile) {
		return this.$fs.deleteFile(fileName);
	} else if (platformInfo && platformInfo.onDeviceName) {
		return this.$fs.rename(fileName, path.join(path.dirname(fileName), platformInfo.onDeviceName));
	}
	return Future.fromResult();
});
Future.wait(futures);
```

Comment (rosen-vladimirov) [inline lib/services/project-files-manager.ts]: after discussion, we've decided to keep the current code

File: lib/services/project-files-manager.ts (hunk 3 above)
Comment (rosen-vladimirov) [inline lib/services/project-files-manager.ts]: after discussion, we've decided to keep the current code

File: lib/services/project-files-manager.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+///<reference path="../.d.ts"/>
+"use strict";
+import path = require("path");
+import util = require("util");
+
+export class ProjectFilesManager implements IProjectFilesManager {
```
Comment (teobugslayer) [inline lib/services/project-files-manager.ts]: Please add unit tests

File: lib/services/plugins-service.ts [hunk 6]
```diff
@@ -24,7 +24,8 @@ export class PluginsService implements IPluginsService {
 		private $options: IOptions,
 		private $logger: ILogger,
 		private $errors: IErrors,
-		private $injector: IInjector) { }
+		private $injector: IInjector,
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: I think `injector` is not used anywhere in this file

File: lib/services/project-files-manager.ts (hunk 5 above)
Comment (Fatme) [inline lib/services/project-files-manager.ts]: I add unit tests.

**Comment by rosen-vladimirov** [comment]
:+1:  after green build

**Comment by teobugslayer** [comment]
:+1:

**Pull request merged**